weave-js/src/components/PagePanelComponents/Home/Browse3/pages/PromptPage/types.ts

```diff
@@ -0,0 +1,31 @@
+// TODO: lock this down
+export type Data = Record<string, any>;
+
+export type Placeholder = {
+  name: string;
+  type: string;
+  default?: string;
+};
+
+export type ImageUrl = {
+  url: string;
+};
+
+export type InternalMessage = {
+  type: 'text' | 'image_url';
+  text?: string;
+  image_url?: ImageUrl;
+};
+
+export type MessagePart = string | Placeholder | InternalMessage;
+
+// TODO: Can we tighten this up?
+export type ToolCall = Record<string, any>;
+
+export type Message = {
+  role: string;
+  content?: string | MessagePart[];
+  tool_calls?: ToolCall[];
+};
+
+export type Messages = Message[];
```

Comment:
`Messages` should be exactly OpenAI's messages. Is that correct?

Reply:
Not exactly the same because I wanted the component to be able to handle our placeholders. I do want to refine these types to get closer to what OpenAI actually has.

Reply:
Can you add a comment on the variation points? due to the union, this appears to be a safe superset. I think the important part is that we can essentially feed openai-compatible outputs into here - which it does imply?